test/test_array_list.cpp

```diff
@@ -28,38 +28,36 @@ class ArrayListTest : public ::testing::Test
 protected:
   void SetUp() override
   {
+    // Reset rcutils error to prevent failure from a previous test case to contaminate
+    // another one.
+    rcutils_reset_error();
+
     allocator = rcutils_get_default_allocator();
     list = rcutils_get_zero_initialized_array_list();
-    rcutils_reset_error();
   }
 
-  // void TearDown() override {}
-
   rcutils_allocator_t allocator;
   rcutils_array_list_t list;
 };
 
 // This fixture is used for test that want the list pre initialized
-class ArrayListPreInitTest : public ::testing::Test
+class ArrayListPreInitTest : public ArrayListTest
 {
 protected:
   void SetUp() override
   {
-    allocator = rcutils_get_default_allocator();
-    list = rcutils_get_zero_initialized_array_list();
-    rcutils_ret_t ret = rcutils_array_list_init(&list, 2, sizeof(uint32_t), &allocator);
+    ArrayListTest::SetUp();
+    const rcutils_ret_t ret = rcutils_array_list_init(&list, 2, sizeof(uint32_t), &allocator);
     EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string().str;
-    rcutils_reset_error();
   }
 
   void TearDown() override
   {
-    rcutils_ret_t ret = rcutils_array_list_fini(&list);
-    EXPECT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string().str;
+    // If the test fails, the test may exit before the list has been initialized
+    // so we allow rcutils_array_list_fini to fail if the list is not initialized.
+    const rcutils_ret_t ret = rcutils_array_list_fini(&list);
+    EXPECT_TRUE(ret == RCUTILS_RET_OK || ret == RCUTILS_RET_NOT_INITIALIZED);
```

Comment:
I still disagree about this, what I'd prefer to see is that you ensure fini is called in any test that does call init. You can accomplish this (as we do in other places in our tests) by using a "scope exit" (there's a header-only implementation provided by `osrf_testing_tools_cpp`), like so:

```c++
rcutils_ret_t ret = rcutils_array_list_init(&list, ...);
ASSERT_EQ(RCUTILS_RET_OK, ret) << rcutils_get_error_string().str;
rcutils_reset_error();
OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT({
  EXPECT_EQ(RCUTILS_RET_OK, rcutils_array_list_fini(&list)) << rcutils_get_error_string().str;
  rcutils_reset_error();
});
```

If you don't reach the scope exit line, the fini will not run, but if you do reach it you know it will get called when the test function exits (for any reason) and you also know that the init was successfully called before it.

Reply:
I understand your suggestion and I applied it in the other PR.

In this case, IIUC your suggestion would mean getting rid of the second pre-existing fixture which is fine but would mean moving back init calls in the test cases everywhere (12 test cases) and as a consequence would expand the scope of this PR. Which strategy would you prefer? Either is fine for me.

Reply:
I think the `ArrayListPreInitTest` is ok (except it should use ASSERT instead in the setup) because it init's in the setup and fini's in the teardown, so they're matched. I don't think it (or the tests that use it) need to be touched.

If any test does not use it, for whatever reason, and does the init themselves then they should use the scope exit to ensure a fini when it is needed.

Reply:
:+1: